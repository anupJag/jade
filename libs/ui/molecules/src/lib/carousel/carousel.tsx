import React, { FunctionComponent, useEffect, useRef, useState } from 'react';
import { Flex } from '@jade/ui/atoms';

import {
  DotControl,
  Slider,
  StyledDot,
  StyledDotsContainer,
  StyledIconButton,
  StyledSlide,
  Track,
} from './carousel.styled';

type CarouselProps = {
  slidesToShow?: number;
  showDots?: boolean;
  showArrow?: boolean;
  children: React.ReactNode;
};

type SlideProps = {
  children: React.ReactNode;
  isActive: boolean;
  isFocused: boolean;
};

type DotsProps = {
  slideCount: number;
  onSlideClick: (value: number) => void;
};

const Dots: FunctionComponent<DotsProps> = ({ slideCount, onSlideClick }) => {
  const dotsArray = Array.from(new Array(slideCount));
  return (
    <StyledDotsContainer>
      <Flex as="ul">
        {dotsArray.map((key, index) => {
          return (
            <StyledDot key={`dot_${index}`}>
              <DotControl
                onClick={() => onSlideClick(index)}
                aria-label={`item ${index + 1} of ${slideCount}`}>
                {' '}
              </DotControl>
            </StyledDot>
          );
        })}
      </Flex>
    </StyledDotsContainer>
  );
};

const Slide: FunctionComponent<SlideProps> = ({ children, isActive, isFocused }) => {
  const slideEl = useRef<HTMLLIElement>(null);

  useEffect(() => {
    if (slideEl && slideEl.current !== null) {
      slideEl.current.querySelectorAll('a, button, input').forEach((element: Element) => {
        element.setAttribute('tabindex', isActive ? '0' : '-1');
      });
    }
  }, [isActive]);

  useEffect(() => {
    if (isFocused && slideEl && slideEl.current !== null) {
      slideEl.current.focus({
        preventScroll: true,
      });
    }
  }, [isFocused]);

  const accessbilityProps = isActive
    ? {
        tabIndex: 0,
      }
    : {
        'aria-hidden': true,
        tabIndex: -1,
      };
  return (
    <StyledSlide ref={slideEl} {...accessbilityProps}>
      {children}
    </StyledSlide>
  );
};

export const Carousel: FunctionComponent<CarouselProps> = ({
  children,
  slidesToShow = 1,
  showDots = true,
  showArrow = true,
}) => {
  const slideCount = React.Children.count(children);
  const [currentSlide, setCurrentSlide] = useState(0);
  const [focusedIndex, setFocusedIndex] = useState(-1);

  const onPrev = () => {
    if (currentSlide > 0) {
      setCurrentSlide(currentSlide - 1);
      setFocusedIndex(currentSlide - 1);
    }
  };

  const onNext = () => {
    if (currentSlide + slidesToShow - 1 < slideCount - 1) {
      setCurrentSlide(currentSlide + 1);
      setFocusedIndex(currentSlide + slidesToShow);
    }
  };

  const setSlide = (index: number) => {
    if (index + slidesToShow < slideCount) {
      setCurrentSlide(index);
    } else {
      setCurrentSlide(slideCount - slidesToShow);
    }
    setFocusedIndex(index);
  };

  return (
    <Slider>
      {showArrow && (
        <StyledIconButton aria-label="prev" icon="chevron-left" variation="prev" onClick={onPrev} />
      )}
      <Track currentSlide={currentSlide} slideCount={slideCount} slidesToShow={slidesToShow}>
        {React.Children.map(children, (child, index) => {
          const isActive = index >= currentSlide && index < currentSlide + slidesToShow; // if slide is in viewport
          const isFocused = focusedIndex > -1 && focusedIndex === index; // if slide needs to be focused after next, prev slide click
          return (
            <Slide key={`slide_${index}`} isActive={isActive} isFocused={isFocused}>
              {child}
            </Slide>
          );
        })}
      </Track>
      {showArrow && (
        <StyledIconButton
          aria-label="next"
          icon="chevron-right"
          variation="next"
          onClick={onNext}
        />
      )}
      {showDots && <Dots slideCount={slideCount} onSlideClick={setSlide} />}
    </Slider>
  );
};
