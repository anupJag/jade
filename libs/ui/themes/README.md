# fe-themes

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `ng test fe-themes` to execute the unit tests via [Jest](https://jestjs.io).
