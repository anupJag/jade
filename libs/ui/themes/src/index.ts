export * from './lib/defaultTheme';
export * from './lib/styled';
export * from './lib/themeProvider';
export { default as styled } from './lib/styled';
export * from './lib/GlobalStyles';
