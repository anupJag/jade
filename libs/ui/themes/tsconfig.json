{
  "extends": "../../../tsconfig.json",
  "compilerOptions": {
    "jsx": "react",
    "allowJs": false,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "types": ["node", "jest"]
  },
  "include": ["**/*.js", "**/*.jsx", "**/*.ts", "**/*.tsx"]
}
