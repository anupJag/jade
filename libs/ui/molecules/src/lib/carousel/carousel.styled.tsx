import { IconButton } from '@jade/ui/atoms';
import { styled } from '@jade/ui/themes';

export const Slider = styled.section`
  overflow: auto;
  width: 100%;
  position: relative;

  ${props => props.theme.mediaQueries.md} {
    overflow: hidden;
  }
`;

export const Track = styled.ul<{
  slideCount: number;
  slidesToShow: number;
  currentSlide: number;
}>`
  display: flex;
  list-style-type: none;
  position: relative;
  width: ${props => `${props.slideCount * 100}%`};

  ${props => props.theme.mediaQueries.md} {
    left: -${props => (props.currentSlide / props.slidesToShow) * 100}%;
    transition: left 600ms;
    width: ${props => (props.slideCount / props.slidesToShow) * 100}%;
  }
`;

export const StyledSlide = styled.li`
  flex-basis: 0;
  flex-grow: 1;
  overflow: hidden;
`;

export const StyledIconButton = styled(IconButton)<{ variation?: string }>`
  box-shadow: 1px 1px 1px #dedede;
  display: none;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  z-index: 1;

  ${props =>
    props.variation === 'next'
      ? `
      right: 0;
      `
      : `
      left: 0;
    `}

  ${props => props.theme.mediaQueries.md} {
    display: block;
  }

  ${props => props.theme.themeColors.cta.primary};
`;

export const StyledDotsContainer = styled.div`
  display: none;
  justify-content: space-around;

  ${props => props.theme.mediaQueries.md} {
    display: flex;
  }
`;

export const StyledDot = styled.li`
  list-style-type: none;
`;

export const DotControl = styled.button`
  border: 1px solid #d3d3d3;
  width: 10px;
  height: 10px;
  min-width: 10px;
  padding: 0;
  border-radius: 5px;
  margin: 0 10px;

  ${props => props.theme.themeColors.cta.primary};
`;

export default StyledIconButton;
