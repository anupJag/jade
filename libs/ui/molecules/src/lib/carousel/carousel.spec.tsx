import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { ThemeProvider } from '@jade/ui/themes';
import '@testing-library/jest-dom/extend-expect';

import { Carousel } from './carousel';

const matchByAriaLabel = label => {
  return (content, element) => element.getAttribute('aria-label') === label;
};

describe('<Carousel />', () => {
  let CarouselComponent;
  beforeEach(() => {
    CarouselComponent = render(
      <ThemeProvider>
        <Carousel>
          {[1, 2, 3, 4, 5].map(slide => (
            <div key={slide}>{slide}</div>
          ))}
        </Carousel>
      </ThemeProvider>,
    );
  });

  test('should render correctly', () => {
    const { asFragment, getByText } = CarouselComponent;
    expect(asFragment()).toMatchSnapshot();
    expect(getByText(matchByAriaLabel('next'))).toBeInTheDocument();
    expect(getByText(matchByAriaLabel('prev'))).toBeInTheDocument();
    expect(getByText(matchByAriaLabel('item 1 of 5'))).toBeInTheDocument();
  });

  test('should contain first slide as visible', () => {
    const { getByRole } = CarouselComponent;
    expect(getByRole('listitem')).toHaveTextContent('1');
  });

  test('on click next, second slide should become visible', () => {
    const { getByRole, getByText } = CarouselComponent;
    fireEvent.click(getByText((content, element) => element.getAttribute('aria-label') === 'next'));
    expect(getByRole('listitem')).toHaveTextContent('2');
  });

  test('on click prev, first slide should become visible if current slide is second', () => {
    const { getByRole, getByText } = CarouselComponent;
    fireEvent.click(getByText(matchByAriaLabel('next')));
    expect(getByRole('listitem')).toHaveTextContent('2');
    fireEvent.click(getByText(matchByAriaLabel('prev')));
    expect(getByRole('listitem')).toHaveTextContent('1');
  });

  test('on click dots, corresponding slide should become visible', () => {
    const { getByRole, getByText } = CarouselComponent;
    fireEvent.click(getByText(matchByAriaLabel('item 3 of 5')));
    expect(getByRole('listitem')).toHaveTextContent('3');
  });
});

describe('Carousel with options', () => {
  test('should not show arrows if showArrow is false', () => {
    const { queryByText } = render(
      <ThemeProvider>
        <Carousel showArrow={false}>
          {[1, 2, 3, 4, 5].map(slide => (
            <div key={slide}>{slide}</div>
          ))}
        </Carousel>
      </ThemeProvider>,
    );

    expect(queryByText(matchByAriaLabel('prev'))).not.toBeInTheDocument();
    expect(queryByText(matchByAriaLabel('next'))).not.toBeInTheDocument();
  });

  test('should not show dots if showDots is false', () => {
    const { queryByText } = render(
      <ThemeProvider>
        <Carousel showDots={false}>
          {[1, 2, 3, 4, 5].map(slide => (
            <div key={slide}>{slide}</div>
          ))}
        </Carousel>
      </ThemeProvider>,
    );

    expect(queryByText(matchByAriaLabel('item 1 in 5'))).not.toBeInTheDocument();
  });

  test('should show 2 slides if slidesToShow is 2', () => {
    const { queryAllByRole } = render(
      <ThemeProvider>
        <Carousel showDots={false} slidesToShow={2}>
          {[1, 2, 3, 4, 5].map(slide => (
            <div key={slide}>{slide}</div>
          ))}
        </Carousel>
      </ThemeProvider>,
    );

    expect(queryAllByRole('listitem')).toHaveLength(2);
  });
});
