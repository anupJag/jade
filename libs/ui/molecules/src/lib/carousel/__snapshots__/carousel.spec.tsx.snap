// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> should render correctly 1`] = `
<DocumentFragment>
  <section
    class="css-13ug728"
  >
    <button
      aria-label="prev"
      class="css-q54x71"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="css-1im46kq"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          fill="currentColor"
        />
      </svg>
    </button>
    <ul
      class="css-nm62o9"
    >
      <li
        class="css-ltvgdk"
        tabindex="0"
      >
        <div>
          1
        </div>
      </li>
      <li
        aria-hidden="true"
        class="css-ltvgdk"
        tabindex="-1"
      >
        <div>
          2
        </div>
      </li>
      <li
        aria-hidden="true"
        class="css-ltvgdk"
        tabindex="-1"
      >
        <div>
          3
        </div>
      </li>
      <li
        aria-hidden="true"
        class="css-ltvgdk"
        tabindex="-1"
      >
        <div>
          4
        </div>
      </li>
      <li
        aria-hidden="true"
        class="css-ltvgdk"
        tabindex="-1"
      >
        <div>
          5
        </div>
      </li>
    </ul>
    <button
      aria-label="next"
      class="css-13t3hkw"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="css-1im46kq"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      class="css-3buig3"
    >
      <ul
        class="css-k008qs"
      >
        <li
          class="css-i6zvb9"
        >
          <button
            aria-label="item 1 of 5"
            class="css-1uqak3w"
          >
             
          </button>
        </li>
        <li
          class="css-i6zvb9"
        >
          <button
            aria-label="item 2 of 5"
            class="css-1uqak3w"
          >
             
          </button>
        </li>
        <li
          class="css-i6zvb9"
        >
          <button
            aria-label="item 3 of 5"
            class="css-1uqak3w"
          >
             
          </button>
        </li>
        <li
          class="css-i6zvb9"
        >
          <button
            aria-label="item 4 of 5"
            class="css-1uqak3w"
          >
             
          </button>
        </li>
        <li
          class="css-i6zvb9"
        >
          <button
            aria-label="item 5 of 5"
            class="css-1uqak3w"
          >
             
          </button>
        </li>
      </ul>
    </div>
  </section>
</DocumentFragment>
`;
